File: src/Rust/backend.rs
use ident::{Ident, Label};
use std::fmt::{Debug, Display};
use tree;

pub trait MachineInstr {
    type UseIterator: Iterator<Item = Ident>;
    type DefIterator: Iterator<Item = Ident>;
    type TargetIterator: Iterator<Item = Label>;

    fn uses(&self) -> Self::UseIterator;

    fn defs(&self) -> Self::DefIterator;

    fn is_fall_through(&self) -> bool;

    fn jumps(&self) -> Self::TargetIterator;

    fn is_move_between_temps(&self) -> Option<(Ident, Ident)>;

    fn is_label(&self) -> Option<Label>;

    fn rename(&mut self, sigma: &Fn(Ident) -> Ident);
}

pub trait MachineFunction<A: MachineInstr>: Debug {
    fn name(&self) -> &Label;
    fn body(&self) -> &Vec<A>;
    fn spill(&mut self, to_spill: &[Ident]);
    fn rename(&mut self, sigma: &Fn(Ident) -> Ident);
}

pub trait MachinePrg<A: MachineInstr, F: MachineFunction<A>>: Debug + Display {
    fn functions(&self) -> &Vec<F>;
    fn functions_mut(&mut self) -> &mut Vec<F>;
}

pub trait Backend {
    type Instr: MachineInstr;
    type Function: MachineFunction<Self::Instr>;
    type Prg: MachinePrg<Self::Instr, Self::Function>;

    fn code_gen(c: &tree::Prg) -> Self::Prg;
}

File: src/Haskell/Backend.hs
{-# LANGUAGE FunctionalDependencies  #-}
module Backend where

import Tree(Prg)
import Names(Temp, Label, MonadNameGen)

class MachineInstr i where
  use  :: i -> [Temp]
  def  :: i -> [Temp]
  jumps :: i -> [Label]
  isFallThrough :: i -> Bool
  isMoveBetweenTemps :: i -> Maybe (Temp, Temp)
  isAssignmentToTemp :: i -> Maybe Temp
  isLabel :: i -> Maybe Label
  renameInstr :: i -> (Temp -> Temp) -> i

class (MachineInstr i) => MachineFunction f i | f -> i where
  machineFunctionName :: f -> String
  machineFunctionBody :: f -> [i]
  machineFunctionRename :: f -> (Temp -> Temp) -> f
  machineFunctionSpill :: MonadNameGen m => f -> [Temp] -> m f

class (MachineInstr i, MachineFunction f i, Show p) => MachinePrg p f i | p -> f i where
  machinePrgFunctions :: p -> [f]
  replaceFunctions :: p -> [f] -> p

class (MachineInstr i, MachineFunction f i, MachinePrg p f i) => CodeGen c p f i | c -> p f i where
  codeGen :: MonadNameGen m => c -> Prg -> m p
  allRegisters :: c -> [Temp]
  generalPurposeRegisters :: c -> [Temp]
